include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -398,9 +397,30 @@ class __future_base
     operator sycl::event() const { return __my_event; }
 };
 
-template <typename T>
+template <typename _T>
 class __future : public __future_base
 {
+    ::std::size_t __result_idx;
+    sycl::buffer<_T> __data;
+    ::std::unique_ptr<__par_backend_hetero::__lifetime_keeper_base> __tmps;
+
+  public:
+    template <typename... _Ts>
+    __future(sycl::event __e, size_t __o, sycl::buffer<_T> __b, _Ts... __t)
+        : __par_backend_hetero::__future_base(__e), __data(__b), __result_idx(__o)
+    {
+        if (sizeof...(_Ts) != 0)
+            __tmps = ::std::unique_ptr<__lifetime_keeper<_Ts...>>(new __lifetime_keeper<_Ts...>(__t...));
+    }
+
+    _T
+    get()
+    {
+        this->wait();
+        return __data.template get_access<access_mode::read>()[__result_idx];
```

Comment:
As we discussed before, the getting an access for whole __data leads to copy the __data to host....  and it is a performance gap. Creating a sub-buffer probably resolves the problem.

Reply:
Good point

Reply:
Since that would be an optimization that doesn't effect the async behavior and could be applied in both cases (blocking and non-blocking), we might consider a new PR.